shared_model/bindings/client_api.cpp

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "bindings/client_api.hpp"
+#include "backend/protobuf/transaction.hpp"
+#include "backend/protobuf/util.hpp"
+#include "common/types.hpp"
+#include "cryptography/crypto_provider/crypto_signer.hpp"
+#include "cryptography/ed25519_sha3_impl/internal/sha3_hash.hpp"
+#include "logger/logger.hpp"
+#include "validators/default_validator.hpp"
+
+namespace shared_model {
+  namespace bindings {
+
+    std::string convert(const Blob &blob) {
+      return std::string(blob.begin(), blob.end());
+    }
+
+    template <typename Proto>
+    boost::optional<Proto> get(const std::string &blob) {
+      Proto proto;
+      if (proto.ParseFromString(blob)) {
+        return proto;
+      }
+      return {};
+    }
+
+    using namespace iroha;
+    std::string validateTransaction(const Blob &b) {
+      auto blob = convert(b);
+      auto s = get<iroha::protocol::Transaction>(blob) | [](auto tx) {
+        static validation::DefaultSignableTransactionValidator val;
+        return boost::make_optional(
+            val.validate(proto::Transaction(tx)).reason());
+      };
+      if (s) {
+        return s.value();
+      }
+      return std::string("unknown object");
+    }
+
+    std::string validateQuery(const Blob &b) {
+      auto blob = convert(b);
+      auto s = get<iroha::protocol::Query>(blob) | [](auto qry) {
+        static validation::DefaultSignableQueryValidator val;
+        return boost::make_optional(val.validate(proto::Query(qry)).reason());
+      };
+      if (s) {
+        return s.value();
+      }
+      return std::string("unknown object");
+    }
+
+    Blob signTransaction(const Blob &b, const crypto::Keypair &key) {
+      auto blob = convert(b);
+      auto s = get<iroha::protocol::Transaction>(blob) | [&key](auto tx) {
+        auto signature =
+            crypto::CryptoSigner<>::sign(proto::makeBlob(tx.payload()), key);
+
+        auto sig = tx.add_signatures();
+        sig->set_signature(crypto::toBinaryString(signature));
+        sig->set_pubkey(crypto::toBinaryString(key.publicKey()));
+        return boost::make_optional(tx);
+      };
+      if (s) {
+        return proto::makeBlob(s.value()).blob();
+      }
+      logger::log("client_api")->warn("Unknown protobuf object");
```

Comment:
what if client does not want to use stdout logger? how to disable it?

I guess there is no way.

IMO all errors should be explicitly mentioned in exceptions

Reply:
2>/dev/null

What kind of exception do you mean?

Reply:
2>/dev/null is not applicable.

Just imagine this situation - you are poor rust developer, you need to use some native library. You write bindings for this library and execute your favorite function. But it starts to write data to STDOUT. 

No good

Reply:
...or you want to disable any logging at all and just catch errors in exceptions.

Reply:
let it be then, I'll just remove